common/flatpak-transaction-log.c

```diff
@@ -0,0 +1,305 @@
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Sam Spilsbury <sam@endlessm.com>
+ */
+
+#include "config.h"
+
+#include <json-glib/json-glib.h>
+
+#include "flatpak-transaction-log.h"
+
+struct FlatpakTransactionLog
+{
+  GObject    parent;
+
+  GFile      *path;
+};
+
+typedef struct
+{
+  GObjectClass parent_class;
+} FlatpakTransactionLogClass;
+
+G_DEFINE_TYPE (FlatpakTransactionLog, flatpak_transaction_log, G_TYPE_OBJECT);
+
+enum {
+  PROP_0,
+  PROP_PATH,
+  LAST_PROP
+};
+
+FlatpakTransactionLog *
+flatpak_transaction_log_new (GFile *path)
+{
+  return g_object_new (FLATPAK_TYPE_TRANSACTION_LOG,
+                       "path", path,
+                       NULL);
+}
+
+static void
+flatpak_transaction_log_finalize (GObject *object)
+{
+  FlatpakTransactionLog *self = (FlatpakTransactionLog *) object;
+
+  g_clear_pointer (&self->path, g_object_unref);
+
+  G_OBJECT_CLASS (flatpak_transaction_log_parent_class)->finalize (object);
+}
+
+static void
+flatpak_transaction_log_get_property (GObject    *object,
+                                      guint       prop_id,
+                                      GValue     *value,
+                                      GParamSpec *pspec)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  switch (prop_id)
+    {
+    case PROP_PATH:
+      g_value_set_object (value, self->path);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+flatpak_transaction_log_set_property (GObject      *object,
+                                      guint         prop_id,
+                                      const GValue *value,
+                                      GParamSpec   *pspec)
+{
+  FlatpakTransactionLog *self = FLATPAK_TRANSACTION_LOG (object);
+
+  switch (prop_id)
+    {
+    case PROP_PATH:
+      g_clear_object (&self->path);
+      self->path = g_value_dup_object (value);
+      break;
+
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static JsonNode *
+load_existing_log_from_file (GFile        *path,
+                             GCancellable *cancellable,
+                             GError       **error)
+{
+  g_autoptr(JsonParser) parser = json_parser_new ();
+  g_autoptr(GFileInputStream) stream = g_file_read (path, cancellable, error);
```

Comment:
You could probably use `json_parser_load_from_file()` instead

Reply:
I had a look at that, though `json_parser_load_from_file` doesn't support cancellability. In any event, we'll need to hand-write a loader for this...